### PR TITLE
adding custom diffusion training to diffusers examples

### PR DESCRIPTION
**Pull request by nupurkmr9**
This PR adds the custom diffusion training script to diffusers examples. [Custom Diffusion](https://arxiv.org/abs/2212.04488) is one of the methods to customize text2image models like stable diffusion given just a few(4~5) images of a subject.
README.md explains the training and inference pipeline using fine-tuned models.

### CODE REVIEW

**Comment by sayakpaul** [comment]
@nupurkmr9 thank you for your contribution! 

I think we need to divide this PR into two (in sequential order):

* The `CustomDiffusionPipeline`. This will include an array of changes. You can refer to https://github.com/huggingface/diffusers/pull/2721 as a reference for how the changes would look like. 
* The training script. 

For all the above contributions, we need to ensure code quality and repository consistency. This is explained here: https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md. 

Let me know if it makes sense or if you have any questions.

**Comment by nupurkmr9** [comment]
Thanks for the references. 
I will make the corresponding changes with only the training script in this PR. And create a new PR for the CustomDiffusionPipeline.

File: examples/custom_diffusion/train.py [hunk 1]
```diff
@@ -0,0 +1,1300 @@
...
+    def change_attn(unet):
+        for layer in unet.children():
+            if type(layer) == CrossAttention:
+                bound_method = set_use_memory_efficient_attention_xformers.__get__(layer, layer.__class__)
+                setattr(layer, 'set_use_memory_efficient_attention_xformers', bound_method)
+            else:
+                change_attn(layer)
+
+    change_attn(unet)
+    unet.set_attn_processor(CustomDiffusionAttnProcessor())
```
Comment (patrickvonplaten) [inline examples/custom_diffusion/train.py:103]: very cool!

File: examples/custom_diffusion/train.py [hunk 2]
```diff
@@ -0,0 +1,1300 @@
...
+
+def freeze_params(params):
+    for param in params:
+        param.requires_grad = False
+
+
+def retrieve(class_prompt, class_images_dir, num_class_images):
+    factor = 1.5
+    num_images = int(factor * num_class_images)
+    client = ClipClient(url="https://knn.laion.ai/knn-service", indice_name="laion_400m", num_images=num_images, aesthetic_weight=0.1)
```
Comment (patrickvonplaten) [inline examples/custom_diffusion/train.py:115]: Could we maybe the user to this as a pre-step and assume that the class images have already been downloaded when starting training?

File: examples/custom_diffusion/train.py [hunk 3]
```diff
@@ -0,0 +1,1300 @@
...
+                    total += 1
+                    pbar.update(1)
+                else:
+                    continue
+            except:
+                continue
+    return
+
+
+def save_model_card(repo_id: str, images=None, base_model=str, prompt=str, repo_folder=None):
```
Comment (patrickvonplaten) [inline examples/custom_diffusion/train.py:155]: super cool!

File: examples/custom_diffusion/train.py [hunk 4]
```diff
@@ -0,0 +1,1300 @@
...
+                    tracker.log(
+                        {
+                            "test": [
+                                wandb.Image(image, caption=f"{i}: {args.validation_prompt}")
+                                for i, image in enumerate(images)
+                            ]
+                        }
+                    )
+
+        if args.push_to_hub:
```
Comment (patrickvonplaten) [inline examples/custom_diffusion/train.py:1279]: cool!

File: examples/custom_diffusion/train.py [hunk 5]
```diff
@@ -0,0 +1,1300 @@
...
+            args.pretrained_model_name_or_path,
+            unet=accelerator.unwrap_model(unet),
+            text_encoder=accelerator.unwrap_model(text_encoder),
+            tokenizer=tokenizer,
+            revision=args.revision,
+            modifier_token=args.modifier_token,
+            modifier_token_id=modifier_token_id,
+        )
+        save_path = os.path.join(args.output_dir, "delta.bin")
+        pipeline.save_pretrained(save_path, freeze_model=args.freeze_model)
```
Comment (patrickvonplaten) [inline examples/custom_diffusion/train.py:1250]: Should we maybe just save the trained parameters instead of the whole pipeline? This would make the output of this training much more portable. E.g. like we do here: https://github.com/huggingface/diffusers/blob/a4b233e5b5092e8ff861b5f5d3ac646fcba9ba79/examples/dreambooth/train_dreambooth_lora.py#L973

File: examples/custom_diffusion/train.py (hunk 2 above)
Comment (nupurkmr9) [reply examples/custom_diffusion/train.py:115]: Yes, I will remove it from the dependency and make it a separate pre-step option.

File: examples/custom_diffusion/train.py (hunk 5 above)
Comment (nupurkmr9) [reply examples/custom_diffusion/train.py:1250]: Yes, currently the CustomDiffusionPipeline saves only trained parameters. But based on later comments, I will re-organize the training script to reuse the existing LoRA and Textual Inversion pipelines. Thanks.

**Comment by sayakpaul** [comment]
Hello @nupurkmr9.

Based on @patrickvonplaten's suggestions, I created this PoC PR (https://github.com/huggingface/diffusers/pull/3088) that might be helpful for showing how we can reuse the existing modules to a good extent. 

Additionally, here's a [notebook](https://gist.github.com/sayakpaul/fd3c0d826705f0f412d6a19164d2ab1a) that shows how the processors can be easily set, serialized, etc. 

Hope that helps :)

**Comment by sayakpaul** [comment]
With @nupurkmr9's hard work, we're in a good position to finally ship this I believe 🚀 🚀 

Custom Diffusion not only supports single-concept training but also multi-concept training very elegantly. Some artifacts below:

* Single concept: [WandB page](https://wandb.ai/sayakpaul/custom-diffusion/runs/26ghrcau), [Hub repo (including results)](https://huggingface.co/sayakpaul/custom-diffusion-cat)
* Multi-concept: [WandB page](https://wandb.ai/sayakpaul/custom-diffusion/runs/3990tzkg), [Hub repo including results](https://huggingface.co/sayakpaul/custom-diffusion-cat-wooden-pot) 

@nupurkmr9 if you could take care of conflict resolution, that would be great.

File: examples/custom_diffusion/README.md [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+image.save("multi-subject.png")
+```
+
+Here, `cat` and `wooden pot` refer to the multiple concepts.
+
+### Inference from a training checkpoint
+
+You can also perform inference from one of the complete checkpoint saved during the training process, if you used the `--checkpointing_steps` argument. 
+
+TODO.
```
Comment (sayakpaul) [inline examples/custom_diffusion/README.md:272]: @patrickvonplaten could you shed some light on how we can do this?

The directory structure of a checkpoint directory is as follows:

image

Kind of confusing as to how to use them.

File: src/diffusers/loaders.py [hunk 7]
```diff
@@ -285,16 +323,31 @@ def save_function(weights, filename):
...
+            model_to_save = AttnProcsLayers(self.attn_processors)
+            state_dict = model_to_save.state_dict()
 
         if weight_name is None:
             if safe_serialization:
-                weight_name = LORA_WEIGHT_NAME_SAFE
+                weight_name = CUSTOM_DIFFUSION_WEIGHT_NAME_SAFE if is_custom_diffusion else LORA_WEIGHT_NAME_SAFE
             else:
-                weight_name = LORA_WEIGHT_NAME
+                weight_name = CUSTOM_DIFFUSION_WEIGHT_NAME if is_custom_diffusion else LORA_WEIGHT_NAME
```
Comment (patrickvonplaten) [inline src/diffusers/loaders.py:350]: Very clean! Great job

File: src/diffusers/models/attention_processor.py [hunk 8]
```diff
@@ -834,4 +1020,6 @@ def __call__(self, attn: "Attention", hidden_states, encoder_hidden_states=None,
     AttnAddedKVProcessor2_0,
     LoRAAttnProcessor,
     LoRAXFormersAttnProcessor,
+    CustomDiffusionAttnProcessor,
+    CustomDiffusionXFormersAttnProcessor,
```
Comment (patrickvonplaten) [inline src/diffusers/models/attention_processor.py:1024]: clean!

File: examples/custom_diffusion/train_custom_diffusion.py [hunk 9]
```diff
@@ -0,0 +1,1281 @@
...
+                truncation=True,
+                padding="max_length",
+                max_length=self.tokenizer.model_max_length,
+                return_tensors="pt",
+            ).input_ids
+
+        return example
+
+
+def save_new_embed(text_encoder, modifier_token_id, accelerator, args, output_dir):
```
Comment (patrickvonplaten) [inline examples/custom_diffusion/train_custom_diffusion.py:293]: Can those be loaded easily after training? Can you load them with: https://huggingface.co/docs/diffusers/api/loaders#diffusers.TextualInversionLoaderMixin.load_textual_inversion?

File: examples/custom_diffusion/train_custom_diffusion.py [hunk 10]
```diff
@@ -0,0 +1,1281 @@
...
+                        )
+
+                del pipeline
+                torch.cuda.empty_cache()
+
+    # Save the custom diffusion layers
+    accelerator.wait_for_everyone()
+    if accelerator.is_main_process:
+        unet = unet.to(torch.float32)
+        unet.save_attn_procs(args.output_dir)
```
Comment (patrickvonplaten) [inline examples/custom_diffusion/train_custom_diffusion.py:1222]: Nice!

File: examples/custom_diffusion/train_custom_diffusion.py [hunk 11]
```diff
@@ -0,0 +1,1281 @@
...
+        pipeline = DiffusionPipeline.from_pretrained(
+            args.pretrained_model_name_or_path, revision=args.revision, torch_dtype=weight_dtype
+        )
+        pipeline.scheduler = DPMSolverMultistepScheduler.from_config(pipeline.scheduler.config)
+        pipeline = pipeline.to(accelerator.device)
+
+        # load attention processors
+        pipeline.unet.load_attn_procs(args.output_dir, weight_name="pytorch_custom_diffusion_weights.bin")
+        for token in args.modifier_token:
+            pipeline.load_textual_inversion(args.output_dir, weight_name=f"{token}.bin")
```
Comment (patrickvonplaten) [inline examples/custom_diffusion/train_custom_diffusion.py:1236]: Ah perfect ok this works for me!

File: examples/custom_diffusion/train_custom_diffusion.py (hunk 9 above)
Comment (patrickvonplaten) [reply examples/custom_diffusion/train_custom_diffusion.py:293]: Ok yes seems to work with: https://github.com/huggingface/diffusers/pull/3031/files#r1169916255

File: examples/custom_diffusion/retrieve.py [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import argparse
+import os
+from io import BytesIO
+from pathlib import Path
+
+import requests
+from clip_retrieval.clip_client import ClipClient
```
Comment (patrickvonplaten) [inline examples/custom_diffusion/retrieve.py:20]: @sayakpaul do yout hink it's ok to leave it here? It's a bit too heavy for me to be a "forced" dependency - see: https://github.com/rom1504/clip-retrieval/blob/main/requirements.txt

File: examples/custom_diffusion/retrieve.py [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+    parser = argparse.ArgumentParser("", add_help=False)
+    parser.add_argument("--class_prompt", help="text prompt to retrieve images", required=True, type=str)
+    parser.add_argument("--class_data_dir", help="path to save images", required=True, type=str)
+    parser.add_argument("--num_class_images", help="number of images to download", default=200, type=int)
+    return parser.parse_args()
+
+
+if __name__ == "__main__":
+    args = parse_args()
+    retrieve(args.class_prompt, args.class_data_dir, args.num_class_images)
```
Comment (patrickvonplaten) [inline examples/custom_diffusion/retrieve.py:87]: Very nice to factor out the retrieval into this file!

File: examples/custom_diffusion/README.md (hunk 6 above)
Comment (patrickvonplaten) [reply examples/custom_diffusion/README.md:272]: Yes this is a good point. To be honest we would need some customized saving logic here to make it easier to run checkpoints in intermediate states, as is done here: https://github.com/huggingface/diffusers/blob/3b641eabe9876e7c48977b35331fda54ce972b4a/examples/dreambooth/train_dreambooth.py#L704

But IMO it's also fine to skip this section in a first PR and only add it at a later stage. The script runs pretty fast and the final checkpoint can be easily loaded no?

Comment (sayakpaul) [reply examples/custom_diffusion/README.md:272]: +1. Totally

**Comment by sayakpaul** [comment]
Happy to help with the tests a bit if needed once I am back. 

Check out the results from https://github.com/huggingface/diffusers/pull/3031#issuecomment-1512597115 and I believe they can be used for the promotion as well.

**Comment by sayakpaul** [comment]
@nupurkmr9 the example tests are failing because of the following.

We're setting the resolution to 64. So, the losses of the individual samples are of the shape: (4, 32, 32).

But the masks are of shape: (1, 8, 8) and hence the mismatch. 

The reason why it doesn't fail with the actual training tests we have conducted is that for 512 resolution, we have the latents of shape (4, 64, 64) and the masks are of shape (1, 64, 64) (sinch 512 / 8 = 64). 

Any ideas to fix this?

**Review by sayakpaul — APPROVED** [review]
Amazing work 🔥 

Thanks so much for adding this and iterating on so much here. Glad that we were able to make use of the existing components and add new things in a coherent manner.

**Comment by patrickvonplaten** [comment]
Great work here!
